* first sample ported from torch tutorials repository

aramaniuk opened the pull request:
the sample is located in samples directory and automatically compiled together with other sources
nginx has configured endpoints for this sample: /impl/samples/1_init  and /impl/samples/1_output 
using the sample: 
$curl http://localhost/impl/samples/1_init
neural network configured
$ curl http://localhost/impl/samples/1_output
basic printing
10
something
number
string
image loaded
image converted from rgb to Y color space

Commit:
* first sample ported from torch tutorials repository

elefthei commented on samples/1_getstarted/output_smpl1.lua:
```diff
@@ -0,0 +1,34 @@
+----------------------------------------------------------------------
+-- snippet 2
+image = require 'image'
+i = image.lena()
```
It would be great to write an image -> tensor converter so we can POST images to it and they get processed

aramaniuk replied:
Totally agree with that. Let me look to this converter stuff
Do you think it should be part of that PR or we have another one?

elefthei commented on samples/1_getstarted/output_smpl1.lua:
```diff
@@ -0,0 +1,34 @@
+ngx.say(a)
+ngx.say 'something'
+ngx.say(type(a))
+ngx.say(type('something'))
```
We don’t need these

aramaniuk replied:
make sense, removed

elefthei commented on nginx/conf/nginx.conf:
```diff
@@ -57,5 +57,15 @@ http {
         }
+
+        # Blocking endpoint
+        location = /impl/samples/1_init {
```
Instead of having one endpoint for each sample algorithm, can we have the Makefile build each sample into a lib/train.lbjc, lib/output.lbjc and then link the right one?

aramaniuk replied:
should this location point to /samples/init, /samples/output or it should replace original init and output

aramaniuk replied:
we can use the global /train and /output endpoints defined in nginx.con and then switch the libraries behind from src to samples
Then we define parameters for Dockerfile to override default init and output with the sample, e.g.
--build-arg activate_sample=1_getstarted

aramaniuk replied:
... or we can specify the sample in runtime by using environment variable
docker run -e activate_sample=1_getstarted -p 3000:3000 -it webtorch

aramaniuk replied:
Please see the latest commit, I have implemented activation of the app in runtime

use parameter activate_app="app_name" where app_name is one of the samples
if activate_app is not provided, then default app is started

docker run -e activate_app="1_getstarted" -p 3000:3000 -it webtorch
Activate app: 1_getstarted

Commit:
merged recent changes from upstream

Commit:
remove useless output

Commit:
dynamically select the application while running container

use parameter activate_app="app_name" where app_name is one of the samples
if activate_app is not provided, then default app is started

elefthei commented on Dockerfile:
```diff
@@ -57,11 +57,18 @@ RUN mkdir /src && \
 ADD . /src/
+
+WORKDIR /src/samples
+RUN make
```
It would be useful to pass an environment variable to make such as 
```
make SRC=/src/samples/1_example
```
Which would build those files and link them as `train.lbjc`, `init.lbjc` and `output.lbjc` when creating the container

aramaniuk replied:
Currently all samples in /src/samples will be compiled producing train.lbjc, init.lbjc and output.lbjc in the folders lib under each sample e.g. /src/samples/{sample_X}/lib
this approach allows to activate any sample without need to rebuild the container. To activate the sample you just pass environment variable activate_app: 
`docker run -e activate_app="1_getstarted" -p 3000:3000 -it webtorch`

elefthei commented on Dockerfile:
```diff
@@ -57,11 +57,18 @@ RUN mkdir /src && \
     cp -r lib /opt/nginx/ && \
     cp nginx/conf/nginx.conf /opt/nginx/conf/nginx.conf
 
+RUN chmod 755 bin/webtorch
```
Better to add it to the above RUN as it will add another layer to the docker image ( a fair amount of space)

aramaniuk replied:
make sense, updated

Commit:
Merge branch 'master' of github.com:elefthei/WebTorch

elefthei reviewed (commented):
Awesome job! A couple of things, instead of specifying the app from the `samples` folder I think it would be even more awesome to allow loading any folder with the following 3 files:
1. init.lua
2. train.lua
3. output.lua

By passing that folder as an env variable to `make` (see comments). This way users can write their own modules fast and just pass them as arguments to make. 

Maybe using docker volumes to mount external folders is a better choice than hardcoding  `/src/samples` in the Dockerfile.

Commit:
Fix automatic test error

Commit:
Unify permission update command into the common RUN

elefthei commented:
Makes sense, if the tests pass let’s merge it.

On Sat, Oct 21, 2017 at 9:16 PM aramaniuk <notifications@github.com> wrote:

> *@aramaniuk* commented on this pull request.
> ------------------------------
>
> In Dockerfile
> <https://github.com/elefthei/WebTorch/pull/1#discussion_r146121462>:
>
> > @@ -57,11 +57,18 @@ RUN mkdir /src && \
>      luarocks install lua-requests
>
>  ADD . /src/
> +
> +WORKDIR /src/samples
> +RUN make
>
> Currently all samples in /src/samples will be compiled producing
> train.lbjc, init.lbjc and output.lbjc in the folders lib under each sample
> e.g. /src/samples/{sample_X}/lib
> this approach allows to activate any sample without need to rebuild the
> container. To activate the sample you just pass environment variable
> activate_app:
>
>
> docker run -e activate_app="1_getstarted" -p 3000:3000 -it webtorch
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/elefthei/WebTorch/pull/1#discussion_r146121462>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAuzrJTbZwHqnJvwO0UYxBpiAcR_HXX3ks5susGEgaJpZM4P5pY5>
> .
>
